### ISSUE TITLE
Question about different outputs from slightly different shapes

### ISSUE THREAD

**Issue by KeunwooPark** [lululxvi/deepxde#141]
Hi.
First of all. DeepXDE is really great!
It is amazing that a neural net can solve PDEs.

I did some experiments and I encountered results that I don't understand. So I hope I can get some explanation about the results.

I'm trying to solve the Laplacian equation for 2D shapes.
The shape has a square with two holes. The boundary-value of the square is 0, and 1,-1 for the two holes.
I wanted to simulate two small disks with opposite charges. 

Here is the code that I used.

```python
from __future__ import absolute_import
from __future__ import division
from __future__ import print_function

import numpy as np

import deepxde as dde
from deepxde.backend import tf

radius = 0.2

def main():
    def pde(x, y):
        dy_x = tf.gradients(y, x)[0]
        dy_x, dy_y = dy_x[:, 0:1], dy_x[:, 1:]
        dy_xx = tf.gradients(dy_x, x)[0][:, 0:1]
        dy_yy = tf.gradients(dy_y, x)[0][:, 1:]
        return -dy_xx - dy_yy

    def boundary(_, on_boundary):
        return on_boundary

    def value_func(x):
        threshold = 0.9
        norm = np.linalg.norm(x, axis=1, keepdims=True)
        val = np.zeros_like(norm)
        val[norm > threshold] = 0
        #val[norm <= threshold] = 1

        val[(x[:,0] < 0) & (x[:,0] > -1) & (x[:,1] < 1) & (x[:,1] > -1)] = 1
        val[(x[:,0] > 0) & (x[:,0] < 1) & (x[:,1] < 1) & (x[:,1] > -1)] = -1
        return val

    bound = dde.geometry.Rectangle([-1,-1], [1,1])
    cathode = dde.geometry.Disk([-0.5,0], radius)
    anode = dde.geometry.Disk([0.5,0], radius)
    geom = dde.geometry.CSGDifference(bound, cathode)
    geom = dde.geometry.CSGDifference(geom, anode)
    bc = dde.DirichletBC(geom, value_func, boundary)

    data = dde.data.PDE(geom, pde, bc, num_domain=1200, num_boundary=120, num_test=15000)
    net = dde.maps.FNN([2] + [50] * 4 + [1], "tanh", "Glorot uniform")
    model = dde.Model(data, net)

    model.compile("adam", lr=0.001)
    losshistory, train_state = model.train(epochs=50000)
    model.compile("L-BFGS-B")
    losshistory, train_state = model.train()
    dde.saveplot(losshistory, train_state, issave=True, isplot=True)


if __name__ == "__main__":
    main()

```

However, the outcomes are very different based on the ```radius``` of the circles.

When ```radius = 0.2```, it shows a reasonable outcome.

![Screenshot from 2020-10-11 13-24-04](https://user-images.githubusercontent.com/1640277/95670584-38d85b00-0bc8-11eb-983f-db7d41639b54.png)
![Screenshot from 2020-10-11 13-24-14](https://user-images.githubusercontent.com/1640277/95670585-3bd34b80-0bc8-11eb-9cef-c683129661dc.png)
![Screenshot from 2020-10-11 13-24-20](https://user-images.githubusercontent.com/1640277/95670590-3e35a580-0bc8-11eb-8f8c-812bc46f5858.png)

On the other hand, when ```radius = 0.1```, it fails to satisfy the boundary conditions.

![Screenshot from 2020-10-11 13-35-01](https://user-images.githubusercontent.com/1640277/95670598-63c2af00-0bc8-11eb-9b31-0aa0b2632590.png)
![Screenshot from 2020-10-11 13-35-11](https://user-images.githubusercontent.com/1640277/95670600-66250900-0bc8-11eb-93d2-acfdce9707f9.png)
![Screenshot from 2020-10-11 13-35-18](https://user-images.githubusercontent.com/1640277/95670601-67eecc80-0bc8-11eb-8a43-8ff7afbb81e8.png)

It is hard for me to interpret this result.
Could someone explain why the two shapes show different results?
Also, I could not find a proper setting that works for ```radius=0.1```. 
Does DeepXDE fail to find an answer when nested boundaries are too small compare to the outermost boundary?

**Comment by lululxvi** [comment]
By looking at the figure of radius=0.1, it seems that at the outside boundary, there are some values to be 1 or -1. I guess it may be due to the following line:
```
val[(x[:,0] < 0) & (x[:,0] > -1) & (x[:,1] < 1) & (x[:,1] > -1)] = 1
val[(x[:,0] > 0) & (x[:,0] < 1) & (x[:,1] < 1) & (x[:,1] > -1)] = -1
```
It is dangerous for float number to do the comparison. For example, you may try to use `x[:, 0]>-0.99` Instead of `x[:, 0]>-1`.

On the other hand, you can plot the locations of training points from `train.dat` to check how many points are in the inside boundary. In total 120 boundary condition might not be enough.

**Comment by KeunwooPark** [comment]
Thank you very much for your advice. 
I updated my code based on your recommendation. 
Moreover, I separated boundary conditions into three and increase the number of boundary points from 120 to 600.

```python
from __future__ import absolute_import
from __future__ import division
from __future__ import print_function

import numpy as np

import deepxde as dde
from deepxde.backend import tf

radius = 0.1

def main():
    def pde(x, y):
        dy_x = tf.gradients(y, x)[0]
        dy_x, dy_y = dy_x[:, 0:1], dy_x[:, 1:]
        dy_xx = tf.gradients(dy_x, x)[0][:, 0:1]
        dy_yy = tf.gradients(dy_y, x)[0][:, 1:]
        return -dy_xx - dy_yy

    def boundary(_, on_boundary):
        return on_boundary

    def boundary_outer(x, on_boundary):
        norm = np.sqrt(x[0]**2 + x[1]**2)
        return on_boundary and (norm > 0.9)

    def boundary_cathode(x, on_boundary):
        return on_boundary & (x[0] < 0) & (x[0] > -0.9) & (x[1] < 0.9) & (x[1] > -0.9)

    def boundary_anode(x, on_boundary):
        return on_boundary & (x[0] > 0) & (x[0] < 0.9) & (x[1] < 0.9) & (x[1] > -0.9)

    def value_outer(x):
        num_data = x.shape[0]
        return np.zeros((num_data,1))

    def value_cathode(x):
        num_data = x.shape[0]
        return np.ones((num_data,1))

    def value_anode(x):
        num_data = x.shape[0]
        return -np.ones((num_data,1))

    bound = dde.geometry.Rectangle([-1,-1], [1,1])
    cathode = dde.geometry.Disk([-0.5,0], radius)
    anode = dde.geometry.Disk([0.5,0], radius)
    geom = dde.geometry.CSGDifference(bound, cathode)
    geom = dde.geometry.CSGDifference(geom, anode)
    bc_outer = dde.DirichletBC(geom, value_outer, boundary_outer)
    bc_cathode = dde.DirichletBC(geom, value_cathode, boundary_cathode)
    bc_anode = dde.DirichletBC(geom, value_anode, boundary_anode)

    data = dde.data.PDE(geom, pde, [bc_outer, bc_cathode, bc_anode], num_domain=1200, num_boundary=600, num_test=15000)
    net = dde.maps.FNN([2] + [50] * 4 + [1], "tanh", "Glorot uniform")
    model = dde.Model(data, net)

    dde.postprocessing.plot_boundary_conditions(data)

    model.compile("adam", lr=0.001)
    losshistory, train_state = model.train(epochs=50000)
    model.compile("L-BFGS-B")
    losshistory, train_state = model.train()
    dde.saveplot(losshistory, train_state, issave=True, isplot=True)


if __name__ == "__main__":
    main()
```
This is how the boundary conditions look like.
![Screenshot from 2020-10-12 13-39-24](https://user-images.githubusercontent.com/1640277/95706024-abad0900-0c90-11eb-8886-072437898a72.png)
![Screenshot from 2020-10-12 13-39-30](https://user-images.githubusercontent.com/1640277/95706028-ae0f6300-0c90-11eb-9a2d-46b4d054ad2e.png)

However, it still could not find the correct answer when ```radius = 0.1```.  It works fine when ```radius = 0.2```and ```radius = 0.15```.
![Screenshot from 2020-10-12 13-54-02](https://user-images.githubusercontent.com/1640277/95707094-8cfc4180-0c93-11eb-8aa6-69df44d337f6.png)
![Screenshot from 2020-10-12 13-54-09](https://user-images.githubusercontent.com/1640277/95707101-91285f00-0c93-11eb-8377-9ad4a9cc3bc0.png)

I don't think decreasing ```radius``` makes the problem any harder. So the same neural network should work. 

Since this is not this DeepXDE's problem, I will not ask further questions. But any advice on this problem will be appreciated.

**Comment by lululxvi** [comment]
You can check FAQ "Q: I failed to train the network or get the right solution, e.g., the training loss is large."

**Comment by KeunwooPark** [comment]
OK. I found the solution.
I had to change the weights for the boundaries. 
This is the working code.
```python
from __future__ import absolute_import
from __future__ import division
from __future__ import print_function

import numpy as np

import deepxde as dde
from deepxde.backend import tf

radius = 0.1

def main():
    def pde(x, y):
        dy_x = tf.gradients(y, x)[0]
        dy_x, dy_y = dy_x[:, 0:1], dy_x[:, 1:]
        dy_xx = tf.gradients(dy_x, x)[0][:, 0:1]
        dy_yy = tf.gradients(dy_y, x)[0][:, 1:]
        return -dy_xx - dy_yy

    def boundary(_, on_boundary):
        return on_boundary

    def boundary_outer(x, on_boundary):
        norm = np.sqrt(x[0]**2 + x[1]**2)
        return on_boundary and (norm > 0.9)

    def boundary_cathode(x, on_boundary):
        return on_boundary & (x[0] < 0) & (x[0] > -0.9) & (x[1] < 0.9) & (x[1] > -0.9)

    def boundary_anode(x, on_boundary):
        return on_boundary & (x[0] > 0) & (x[0] < 0.9) & (x[1] < 0.9) & (x[1] > -0.9)

    def value_outer(x):
        num_data = x.shape[0]
        return np.zeros((num_data,1))

    def value_cathode(x):
        num_data = x.shape[0]
        return np.ones((num_data,1)) * 1

    def value_anode(x):
        num_data = x.shape[0]
        return -np.ones((num_data,1)) * 1

    bound = dde.geometry.Rectangle([-1,-1], [1,1])
    cathode = dde.geometry.Disk([-0.5,0], radius)
    anode = dde.geometry.Disk([0.5,0], radius)
    geom = dde.geometry.CSGDifference(bound, cathode)
    geom = dde.geometry.CSGDifference(geom, anode)
    bc_outer = dde.DirichletBC(geom, value_outer, boundary_outer)
    bc_cathode = dde.DirichletBC(geom, value_cathode, boundary_cathode)
    bc_anode = dde.DirichletBC(geom, value_anode, boundary_anode)

    data = dde.data.PDE(geom, pde, [bc_outer, bc_cathode, bc_anode], num_domain=1200, num_boundary=600, num_test=15000)
    net = dde.maps.FNN([2] + [50] * 4 + [1], "tanh", "Glorot uniform")
    model = dde.Model(data, net)

    dde.postprocessing.plot_boundary_conditions(data)

    model.compile("adam", lr=0.001, loss_weights = [1, 100, 10, 10])
    losshistory, train_state = model.train(epochs=50000)
    model.compile("L-BFGS-B")
    losshistory, train_state = model.train()
    dde.saveplot(losshistory, train_state, issave=True, isplot=True)


if __name__ == "__main__":
    main()
```

And these are the results.
![Screenshot from 2020-10-13 12-24-18](https://user-images.githubusercontent.com/1640277/95811896-3a805b00-0d4f-11eb-8b91-6b4b1aa6df20.png)
![Screenshot from 2020-10-13 12-24-25](https://user-images.githubusercontent.com/1640277/95811899-3ce2b500-0d4f-11eb-946e-19f5352ff302.png)
![Screenshot from 2020-10-13 12-24-33](https://user-images.githubusercontent.com/1640277/95811906-3f450f00-0d4f-11eb-8ae2-5ef8bd1d7832.png)

The key point is to set large weights for the outer boundaries and small weights for the inner boundaries. And PDE loss should the the smallest.

**Comment by katayooneshkofti** [comment]
@KeunwooPark Thank you for the tips
